File: src/day14/main.exs
defmodule Day14 do
  @moduledoc "AoC2022 Day14 functions"

  @spec drop_sand(map, atom, integer) :: {map, integer}
  def drop_sand(blocked, endcond, endy, x \\ 500, y \\ 0) do
    if Map.has_key?(blocked, {x, y}) do
      nil
    else
      case endcond do
        :floor when y >= endy ->
          {blocked, {x, y}}

        :void when y >= endy ->
          nil

        _ ->
          next =
            Enum.find([x, x - 1, x + 1], nil, fn x ->
              !Map.has_key?(blocked, {x, y + 1})
            end)

          case next do
            nil -> {blocked, {x, y}}
            x -> drop_sand(blocked, endcond, endy, x, y + 1)
          end
      end
    end
  end

  def play(blocked, endcond, endy, count \\ 0) do
    case drop_sand(blocked, endcond, endy) do
      nil ->
        {blocked, count}

      {blocked, key} ->
        play(Map.put(blocked, key, :sand), endcond, endy, count + 1)
    end
  end
end

blocked =
  for line <-
        File.read!("puzzle_input/day14")
        |> String.trim_trailing()
        |> String.split(~r/\R/, trim: True) do
    coords =
      String.split(line, " -> ")
      |> Enum.map(fn s ->
        String.split(s, ",")
        |> Enum.map(fn s -> String.to_integer(s) end)
      end)

    for {[x1, y1], [x2, y2]} <- Stream.zip(coords, Stream.drop(coords, 1)) do
      {x1, x2} =
        if x1 <= x2 do
          {x1, x2}
        else
          {x2, x1}
        end

      {y1, y2} =
        if y1 <= y2 do
          {y1, y2}
        else
          {y2, y1}
        end

      for y <- y1..y2, x <- x1..x2 do
        {{x, y}, :rock}
      end
    end
  end
  |> Stream.concat()
  |> Stream.concat()
  |> Map.new()

{{_, lowest}, _} = Enum.max(blocked, fn {{_, y1}, _}, {{_, y2}, _} -> y1 >= y2 end)

{blocked, p1} = Day14.play(blocked, :void, lowest + 1)
{_, p2} = Day14.play(blocked, :floor, lowest + 12)

# 913, 30762
IO.puts("part 1: #{p1}\npart 2: #{p2}")

File: src/day14/main.jl
# julia src/day14/main.jl
EXAMPLE = "498,4 -> 498,6 -> 496,6
503,4 -> 502,4 -> 502,9 -> 494,9"
ROCK = 0
SAND = 1

const Coord = Tuple{Int64,Int64}
const State = Dict{Coord,Int64}

function build_map()::State
    blocked = State()
    for line in readlines("puzzle_input/day14")
        coords = split(line, " -> ") .|>
                 (raw_coord -> begin
            split(raw_coord, ",") .|>
            (s -> parse(Int64, s))
        end)
        for ((x1, y1), (x2, y2)) in zip(coords, coords[2:end])
            x1, x2 = minmax(x1, x2)
            y1, y2 = minmax(y1, y2)
            for coord in Iterators.product(x1:x2, y1:y2)
                blocked[coord] = ROCK
            end
        end
    end
    blocked
end

function draw(blocked)
    left, right, top, bottom = 500, 500, 0, 0

    for (x, y) in keys(blocked)
        left = min(left, x)
        right = max(right, x)
        top = min(top, y)
        bottom = max(bottom, y)
    end

    left -= 1
    right += 1
    bottom += 1

    for y in top:bottom
        for x in left:right
            if haskey(blocked, (x, y))
                type = blocked[(x, y)]
                if type == ROCK
                    print("%%")
                else
                    print("::")
                end
            else
                print("  ")
            end
        end
        println()
    end
end

function find(f, xs)
    for x in xs
        if f(x)
            return x
        end
    end
    nothing
end

function drop_sand(blocked::State, has_floor::Bool, bottom::Int64)::Union{Coord,Nothing}
    x = 500
    y = 0
    if haskey(blocked, (x, y))
        return nothing
    end
    while true
        # return once void reached, or come to rest on floor
        if y >= bottom
            return has_floor ? (x, y) : nothing
        end
        # fall one step
        next = find([x, x - 1, x + 1]) do x
            !haskey(blocked, (x, y + 1))
        end
        if nothing === next
            # found where it came to rest!
            return (x, y)
        end
        x = next
        y = y + 1
    end
end

function play(blocked, has_floor, bottom)
    while true
        coord = drop_sand(blocked, has_floor, bottom)
        if nothing === coord
            break
        end
        blocked[coord] = SAND
    end
end

blocked = build_map()

lowest_rock = 0
for (_, y) in keys(blocked)
    global lowest_rock = max(lowest_rock, y)
end

rock_count = length(blocked)

play(blocked, false, lowest_rock)
#draw(blocked)
println("part 1: ", length(blocked) - rock_count) # 913

play(blocked, true, lowest_rock + 1)
#draw(blocked)
println("part 2: ", length(blocked) - rock_count) # 30762